File: solutions/lua/bottle-song/1/bottle-song.lua
local BottleSong = {}

local verse = [[
BEFORE hanging on the wall,
BEFORE hanging on the wall,
And if one green bottle should accidentally fall,
There'll be AFTER hanging on the wall.
]]

local numbers = { "one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "Ten" }

local function bottles(n)
	if n == 1 then
		return "one green bottle"
	elseif n == 0 then
		return "no green bottles"
	else
		return numbers[n] .. " green bottles"
	end
end

function BottleSong.recite(start_bottles, take_down)
	local song = {}

	for i = start_bottles, start_bottles - take_down + 1, -1 do
		local before = bottles(i):gsub("^%l", string.upper)
		local after = bottles(i - 1)
		local this = verse:gsub("BEFORE", before):gsub("AFTER", after)
		table.insert(song, this)
	end

	return table.concat(song, "\n")
end

return BottleSong

File: solutions/elixir/bottle-song/1/lib/bottle_song.ex
defmodule BottleSong do
  @moduledoc """
  Handles lyrics of the popular children song: Ten Green Bottles
  """

  @spec recite(pos_integer, pos_integer) :: String.t()
  def recite(start_bottle, take_down) do
    verses =
      for n <- start_bottle..(start_bottle - take_down + 1)//-1 do
        at_start = n |> bottles() |> String.capitalize()

        """
        #{at_start} hanging on the wall,
        #{at_start} hanging on the wall,
        And if one green bottle should accidentally fall,
        There'll be #{bottles(n - 1)} hanging on the wall.\
        """
      end

    Enum.join(verses, "\n\n")
  end

  @words ~w(no one two three four five six seven eight nine ten)

  defp bottles(1), do: "one green bottle"
  defp bottles(n), do: "#{Enum.at(@words, n)} green bottles"
end
